ocaml/xapi-idl/xen/xenops_types.ml

```diff
@@ -129,7 +129,11 @@ module Vm = struct
   }
   [@@deriving rpcty, sexp]
 
-  type builder_info = HVM of hvm_info | PV of pv_info | PVinPVH of pv_info
+  type builder_info =
+    | HVM of hvm_info
+    | PV of pv_info
+    | PVinPVH of pv_info
+    | PVH of pv_info
   [@@deriving rpcty, sexp]
```

Comment:
At some point we're going to have to have a conversation about how Xapi's data model ought to change...

Reply:
I suspect pv_info is not entirely accurate here now and we'd want a mix of hvm_info and pv_info?

Reply:
I don't think a suitable data model would be expressed in those terms to begin with

Reply:
maybe it'd be better to express this in terms of devices or components and their configurations. There are some implicit assumptions leaking through here, e.g. 'tpm' is only present in hvm_info because it is currently implemented by qemu, and with no qemu there is no tpm for pvh. If we'd split the bits that are specific to configuring Xen from the bits specific to configuring QEMU, and the bits from configuring devices that are usable by either pv/pvh/hvm that might be better.

One possibility here is to use phantom types, e.g. 'tpm' could be an `'Swtpm_qemu` device because that is how it is currently implemented, and HVM could accept [Swtpm_qemu | Qemu | HVM | Common] configs). Similarly 'acpi' could be an 'HVM' config, so it won't get accepted by PV, but get accepted by PVH (probably?).
And for pvinpvh we need to differentiate arguments given to nested Xen too.